lib/AST/Type.cpp

```diff
@@ -3240,12 +3240,29 @@ TypeSubstitutionMap TypeBase::getContextSubstitutions(const DeclContext *dc) {
 
 SubstitutionMap TypeBase::getContextSubstitutionMap(
   ModuleDecl *module, const DeclContext *dc) {
+  auto lookup = LookUpConformanceInModule(module);
+
+  // If we're substituting in a protocol, we can avoid the full
+  // generic-signature-based lookup, because the substitutions are easy to
+  // compute. It is important to do this, because name-lookup needs to use this,
+  // and the generic signature of a protocol may not yet be computed.
```

Comment:
The signature itself should be trivial `<Self: Proto>`; when we create a `GenericSignatureBuilder` we'll explode out the implied requirements so they can be queried.

To be fair, our nascent plans for storing concrete conformances into the signature will mean that `<Self: Proto>` won't always be trivial. We'll have to think about how to handle that... later.